apis/v1alpha1/shared_types.go

```diff
@@ -175,6 +189,31 @@ type AdminNetworkPolicyEgressPeer struct {
 	// <network-policy-api:experimental>
 	// +optional
 	Nodes *metav1.LabelSelector `json:"nodes,omitempty"`
+	// Networks defines a way to select peers via CIDR blocks.
+	// This is intended for representing entities that live outside the cluster,
+	// which can't be selected by pods, namespaces and nodes peers, but note
+	// that cluster-internal traffic will be checked against the rule as
+	// well. So if you Allow or Deny traffic to `"0.0.0.0/0"`, that will allow
```

Comment:
> cluster-internal traffic will be checked against the [CIDR] rule as well. 

Cilium doesn't work this way today. And I think we have reasonable grounds not to work this way - there are better representations for handing policy for Pods, and we make network policy decisions based on the security identity of a peer, not its address. Differentiating by individual Pod addresses would expand out Cilium's policy representation at the datapath layer, leading to higher memory usage and likely higher CPU cost. Unless there's a strong user-motivated reason to support this, I'd rather not set the expectation it will work universally.

For a standard, if we intend to accommodate a range of behaviour, I think the best we can say is that "The ability for CIDR blocks to select traffic corresponding to a Pod IP is implementation-defined".

Reply:
yeah we spoke about this a little bit on our meeting this week and Nathan mentioned Cilium doesn't allow or count podCIDRs as a CIDR and you'd need to use pod selectors and that I think is true for an ideal world and recommended but for the case of syntax accuracy when we say 0.0.0.0/0 ultimately its everything, so we spent a lot of time discussing that point the last few months and ended up with this consensus that cluster-internal traffic will also be checked because that is ultimately easier for implementations. We definitely didn't want to do "implementation defined" because we wanted there to be some consistency in how implementations did it and not have the same mistake like with NetPol where everybody did different things.

Reply:
started a new PR to track this work: https://github.com/kubernetes-sigs/network-policy-api/pull/204

Reply:
Is it crazy to explore the idea of a definition like `type ExternalNetworks []CIDR`? In all of the ipBlock / CIDR / Networks text I'm aware of, they all start with this expression that "This is intended for representing entities that live outside the cluster", so if that's the intent, then maybe that semantic should be embedded into the field naming...? I must admit I struggle a little bit with the idea that everyone agrees this shouldn't be used this for internal/pod-to-pod traffic, but we're trying to solve that use case anyway.

In the back of my mind, I'm also thinking a lot about the idea that network policy should be trying to specify _who_ one is trying to talk to, ideally not _where_. One of the great pushes forward with cloud native & k8s is that with the high Pod churn, IP address reuse, etc. we are largely _forced_ to adopt richer notions of who can talk to who - which objects can talk to which other objects, not which locations can talk to other locations (regardless of the applications running at those locations). Cilium project features like ToFQDNs demonstrate a strong desire to specify the policy based on higher level concepts of the services being contacted rather encoding a bunch of addresses in the policies (which are hard to understand and harder to update).

> Option 1 doesn't make sense with ANP's syntax; you can have a single ANP containing a cidr rule nested between two pod rules; it would be confusing for that to be creating rules in two totally separate priority ranges. 

Could you expand this example to demonstrate the aspect that doesn't make sense? I don't understand the "totally separate priority ranges" aspect of this. We know that Pod selectors will never select traffic outside the cluster, so then the remaining question is whether Networks selectors intersect with those Pod selectors or not. Put another way, is the idea that you have

Reply:
> Is it crazy to explore the idea of a definition like `type ExternalNetworks []CIDR`? In all of the ipBlock / CIDR / Networks text I'm aware of, they all start with this expression that "This is intended for representing entities that live outside the cluster", so if that's the intent, then maybe that semantic should be embedded into the field naming...? I must admit I struggle a little bit with the idea that everyone agrees this shouldn't be used this for internal/pod-to-pod traffic, but we're trying to solve that use case anyway.

yeah this was extensively discussed multiple times in the net-pol meetings and we even went back and forth a bit before landing on this. This was actually called the `externalNetworks` int he first iteration of the PR and then moved to being `networks` because what does 0.0.0.0/0 in `externalNetworks` mean?
I think the other NPEP I have planned for the future where we want to do a new object representing Networks.. there I might go this way of making it only externalNetworks.

Reply:
TBH at this point to satisfy all use cases I might actually do `networks` and `externalNetworks` where former includes evaluating all CIDRs and latter excludes evaluating podCIDRs and nodeCIDRs a.ka intra cluster traffc? Because there are other implementations who have mentioned its actually hard to "define what is external to a cluster" that is another reason why we did pivot towards this in this manner.

Reply:
anyways I have tried to capture all of this: https://docs.google.com/presentation/d/1gAZq15nVavhLR29ZfqXzrUAA3Vy2HIHD_IvxyZ0StME/edit?usp=sharing
 it really boils down to that, if needed I can add a paragraph to the existing NPEP but right now the NPEP is moving towards experimental phase meaning these points were already discussed extensively while I am unsure on how to satisfy all implementations here.

Reply:
we discussed this on slack; https://kubernetes.slack.com/archives/C01CWSHQWQJ/p1709482466315219 resolving this conversation.